### ISSUE TITLE
Potentially Missing labels and modality pairs

### ISSUE THREAD

**Issue by nilsleh** [aysim/dynnet#10]
Hi, 

I am working on adding the dataset to torchgeo, but I am running into an issue, where it seems that the labels.zip file from the TUM server is incomplete.

I have used the following code to check whether the labels path exists in the labels directory for each of the split text files as such:

```python
import pandas as pd
import os

splits = ['train', 'val', 'test']

df = pd.DataFrame(columns=['split', 'image_path', 'label_path', 'year_month'])
for split in splits:
    with open(f'{split}.txt', 'r') as f:
        lines = f.readlines()

    # Each line has three parts: "folder path", "tif_path", and "date"
    parsed_data = [line.strip().split(" ") for line in lines]
    split_df = pd.DataFrame(parsed_data, columns=['image_path', 'label_path', 'year_month'])

    # check if the label path exists
    
    missing = []
    for i, row in split_df.iterrows():
        path = os.path.join(os.path.dirname(os.getcwd()), row['label_path'].lstrip('/'))
        if not os.path.exists(path):
            missing.append(os.path.basename(os.path.dirname(row['label_path'])))

    print(split)
    print(set(missing))
```

And get the following output.

```python
train
{'38S-44E-19S-L3H-SR', '52N-128E-36N-L3H-SR', '46N-92E-25N-L3H-SR', '54S-143E-38S-L3H-SR', '19S-71W-13S-L3H-SR'}
val
{'35N-25E-60N-L3H-SR', '21S-57W-25S-L3H-SR', '47N-99E-18N-L3H-SR', '19S-67W-18S-L3H-SR', '50N-117E-39N-L3H-SR', '43N-76E-11N-L3H-SR', '56S-150E-33S-L3H-SR', '30N-1W-6N-L3H-SR', '15N-93W-16N-L3H-SR', '49N-113E-22N-L3H-SR'}
test
{'44N-79E-6N-L3H-SR', '29N-11W-7N-L3H-SR', '34S-18E-5S-L3H-SR', '43N-72E-21N-L3H-SR', '10N-123W-45N-L3H-SR', '39N-48E-29N-L3H-SR', '10N-121W-39N-L3H-SR', '34N-20E-32N-L3H-SR', '17N-83W-9N-L3H-SR', '50N-116E-39N-L3H-SR'}
```

I have verified that the sha512sum of the downloaded labels.zip file is the same. So either there are some labels missing, or the `.txt` files are not correct? Maybe you could verify this on your side? 

There seem to be 120 labels missing for train, and then all labels for validation and test.

@aysim @lukaskondmann

**Comment by nilsleh** [comment]
There also seems to be some sort of mismatch with the additional sentinel 1 and 2 data.

For sentinel1:

```python
missing_s1
['sentinel1/1487_3335_13/1487_3335_13_2019_08.tif', 'sentinel1/2415_3082_13/2415_3082_13_2019_04.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_02.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_03.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_04.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_05.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_06.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_07.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_08.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_09.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_10.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_11.tif', 'sentinel1/2624_4314_13/2624_4314_13_2018_12.tif', 'sentinel1/2624_4314_13/2624_4314_13_2019_01.tif', 'sentinel1/2624_4314_13/2624_4314_13_2019_02.tif', 'sentinel1/2624_4314_13/2624_4314_13_2019_03.tif', 'sentinel1/2624_4314_13/2624_4314_13_2019_04.tif', 'sentinel1/2624_4314_13/2624_4314_13_2019_05.tif', 'sentinel1/2624_4314_13/2624_4314_13_2019_06.tif', 'sentinel1/2624_4314_13/2624_4314_13_2019_07.tif', 'sentinel1/3002_4273_13/3002_4273_13_2019_10.tif', 'sentinel1/4240_3972_13/4240_3972_13_2018_01.tif', 'sentinel1/4240_3972_13/4240_3972_13_2018_02.tif', 'sentinel1/4856_4087_13/4856_4087_13_2018_01.tif', 'sentinel1/4856_4087_13/4856_4087_13_2018_02.tif', 'sentinel1/4856_4087_13/4856_4087_13_2018_03.tif', 'sentinel1/4856_4087_13/4856_4087_13_2018_04.tif', 'sentinel1/4856_4087_13/4856_4087_13_2018_05.tif', 'sentinel1/6824_4117_13/6824_4117_13_2018_07.tif', 'sentinel1/6824_4117_13/6824_4117_13_2018_08.tif', 'sentinel1/6824_4117_13/6824_4117_13_2018_09.tif']
```

For sentinel2:

```python
['sentinel2/2029_3764_13/2018_01.tif', 'sentinel2/2029_3764_13/2018_02.tif', 'sentinel2/2029_3764_13/2018_03.tif', 'sentinel2/2029_3764_13/2018_04.tif', 'sentinel2/2029_3764_13/2018_05.tif', 'sentinel2/2029_3764_13/2018_06.tif', 'sentinel2/2029_3764_13/2018_07.tif', 'sentinel2/2029_3764_13/2018_08.tif', 'sentinel2/2029_3764_13/2018_09.tif', 'sentinel2/2029_3764_13/2018_10.tif', 'sentinel2/2029_3764_13/2018_11.tif', 'sentinel2/2029_3764_13/2018_12.tif', 'sentinel2/2029_3764_13/2019_01.tif', 'sentinel2/2029_3764_13/2019_02.tif', 'sentinel2/2029_3764_13/2019_03.tif', 'sentinel2/2029_3764_13/2019_04.tif', 'sentinel2/2029_3764_13/2019_05.tif', 'sentinel2/2029_3764_13/2019_06.tif', 'sentinel2/2029_3764_13/2019_07.tif', 'sentinel2/2029_3764_13/2019_08.tif', 'sentinel2/2029_3764_13/2019_09.tif', 'sentinel2/2029_3764_13/2019_10.tif', 'sentinel2/2029_3764_13/2019_11.tif', 'sentinel2/2029_3764_13/2019_12.tif', 'sentinel2/4169_3944_13/2019_07.tif', 'sentinel2/6381_3681_13/2019_08.tif', 'sentinel2/6466_3380_13/2018_10.tif', 'sentinel2/6466_3380_13/2019_11.tif', 'sentinel2/6475_3361_13/2019_11.tif', 'sentinel2/6688_3456_13/2019_01.tif', 'sentinel2/6810_3478_13/2019_02.tif']
```

**Comment by aysim** [comment]
Hi,

Sorry for the late response. The labels are not provided for the test and validation sets, which is why they are missing.

Regarding the training set, there may be slight discrepancies between the naming of my splits and the original TUM dataset. For example, in my training splits, the cube name is "5111_4560_13_38S" for "38S-44E-19S-L3H-SR", whereas in the TUM dataset folder, the name might appear as "5111_4560_13-38S"—with a difference in the use of "_" and "-".

Could you please double-check and adjust your code accordingly?

**Comment by nilsleh** [comment]
Thanks for your reply, indeed that is the issue for several directories:

```
/labels/5111_4560_13_38S 
/labels/6204_3495_13_46N
/labels/7026_3201_13_52N
/labels/7367_5050_13_54S
/labels/2459_4406_13_19S
/labels/2235_3403_13_17N
```

will rehost the data on hugginface with corrected directories.

**Comment by nilsleh** [comment]
@aysim  I have a follow up question, since it seems that for some training entries, based on the `train.txt` file some sentinel1 or sentinel2 scenes are missing. I have made a script to check the presence of all the modality parts, as well as adding new entries according to the missing event_ids you noted in the README. 

```python
import pandas as pd
import os
from glob import glob

def check_file_exists(path):
    return os.path.exists(os.path.join(os.path.dirname(os.getcwd()), path.lstrip('/')))

def process_split_file(split):
    with open(f'{split}.txt', 'r') as f:
        parsed_data = [line.strip().split(" ") for line in f.readlines()]
        
    df = pd.DataFrame(parsed_data, columns=['planet_path', 'label_path', 'year_month'])
    df['split'] = split
    
    missing_labels = [i for i, row in df.iterrows() if not check_file_exists(row['label_path'])]
    df['missing_label'] = False
    df.loc[missing_labels, 'missing_label'] = True
    
    df['s1_path'] = df.apply(
        lambda row: f"sentinel1/{row['planet_path'].split('/')[3]}/{row['planet_path'].split('/')[3]}_{row['year_month'].replace('-', '_')}.tif",
        axis=1
    )
    df['s2_path'] = df.apply(
        lambda row: f"sentinel2/{row['planet_path'].split('/')[3]}/{row['year_month'].replace('-', '_')}.tif",
        axis=1
    )
    
    missing_s1 = [i for i, row in df.iterrows() if not check_file_exists(row['s1_path'])]
    missing_s2 = [i for i, row in df.iterrows() if not check_file_exists(row['s2_path'])]
    
    df['missing_s1'] = False
    df['missing_s2'] = False
    df.loc[missing_s1, 'missing_s1'] = True
    df.loc[missing_s2, 'missing_s2'] = True
    
    df['planet_path'] = df['planet_path'].apply(lambda x: f"planet{x}")
    df['label_path'] = df['label_path'].apply(lambda x: x.lstrip('/'))
    
    return df

def process_additional_entries(split):
    new_entries = ['1417_3281_13_11N', '2235_3403_13_17N']
    result_df = pd.DataFrame(columns=['split', 'label_path'])
    
    for entry in new_entries:
        path = os.path.join(os.path.dirname(os.getcwd()), f"labels/{entry}/Raster")
        more_labels = sorted(glob(os.path.join(path, "**", "*.tif"), recursive=True))
        result_df = pd.concat([result_df, pd.DataFrame({'split': split, 'label_path': more_labels})], ignore_index=True)
    
    result_df["year_month"] = result_df["label_path"].apply(
        lambda x: x.split("/")[-1].split("-")[-1].split(".")[0]
    ).str.replace("_", "-").str[:7]
    
    result_df['label_path'] = result_df['label_path'].apply(
        lambda x: x.replace(os.path.dirname(os.getcwd()), "").lstrip("/")
    )
    
    result_df["planet_path"] = result_df["label_path"].apply(
        lambda x: '/'.join([
            x.split('/')[1].split('_')[-1],
            '-'.join(x.split('/')[-1].split('-')[1:3]) if len(x.split('/')[-1].split('-')) >= 3 else "unknown-unknown",
            '_'.join(x.split('/')[1].split('_')[:-1]),
            "PF-SR"
        ])
    )
    
    result_df['planet_path'] = result_df['planet_path'].apply(lambda x: f"planet{x}")
    
    result_df["missing_label"] = False
    result_df["missing_s1"] = False
    result_df["missing_s2"] = False
    
    result_df['s1_path'] = result_df.apply(
        lambda row: f"sentinel1/{row['planet_path'].split('/')[-2]}/{row['planet_path'].split('/')[-2]}_{row['year_month'].replace('-', '_')}.tif",
        axis=1
    )
    result_df['s2_path'] = result_df.apply(
        lambda row: f"sentinel2/{row['planet_path'].split('/')[-2]}/{row['year_month'].replace('-', '_')}.tif",
        axis=1
    )
    
    missing_s1 = [i for i, row in result_df.iterrows() 
                if not check_file_exists(row['s1_path'])]
    missing_s2 = [i for i, row in result_df.iterrows() 
                if not check_file_exists(row['s2_path'])]
    
    result_df.loc[missing_s1, 'missing_s1'] = True
    result_df.loc[missing_s2, 'missing_s2'] = True
    
    return result_df

if __name__ == "__main__":
    splits = ['train']
    main_df = pd.DataFrame(columns=['split', 'planet_path', 'label_path', 'year_month'])
    
    for split in splits:
        split_df = process_split_file(split)
        main_df = pd.concat([main_df, split_df], ignore_index=True)
        
        if split == "train":
            additional_df = process_additional_entries(split)
            main_df = pd.concat([main_df, additional_df], ignore_index=True)
    
    main_df.to_csv('splits.csv', index=False)
```

If I use this script from the root directory of the datasets, then I get 1296 total rows, and for 50 of them the sentinel 1 scene is missing in the dataset, and for 32 of them the sentinel 2 scene is missing. I have uploaded the resulting csv here: https://gist.github.com/nilsleh/4eb593faed6a2290e6b23adda0cb3d0a

I have redownloaded and unzipeed the sentinel 1 directory, but those files are missing, for example, here, the `1487_3335_13/1487_3335_13_2019_08.tif` seems to be missing.

![Image](https://github.com/user-attachments/assets/7fbfd9d2-98f9-4f29-907b-156ce9969ad0)

**Comment by aysim** [comment]
Since obtaining monthly data from Sentinel-1 and Sentinel-2 is challenging for various reasons, this might be contributing to the issue. In our benchmarking, we primarily used Planet imagery but included Sentinel-1 and Sentinel-2 to support further research. For the huggingface dataloader, one option could be to skip entries without Sentinel-1 and Sentinel-2 data, or alternatively, you could use the closest available date to fill the gaps.

**Comment by nilsleh** [comment]
Yes, that would be a strategy, however, I am more wondering whether the uploaded dataset is complete or not. In A.2 of the paper you mention that you have created monthly composites, which points out potential quality issues, but does not mention that data would be missing.

**Comment by aysim** [comment]
it's possible that all images from that specific month have quality issues. The uploaded data is the final version, so we’ll have to work with it as is.

**Comment by nilsleh** [comment]
@aysim Sorry, last question. The sentinel 1 tiff files have 8 bands, but I could not find which bands those actually are. Similarly for Sentinel 2, there are 12 bands but I could not find documentation or info which bands are included.

**Comment by nilsleh** [comment]
@aysim Sorry, I lied one more additional question:

In Appendix A.3 the different modality settings are described:

- "For the weekly setting, we feed the architectures with samples from the 1st, 5th, 10th, 15th, 20th and 25th days of each month"

I am a bit confused by the temporal difference between the inputs and the target label. Let's say in the weekly setting , we are in month "2018-01". When I step through the dataset [here](https://github.com/aysim/dynnet/blob/1e7d90294b54f52744ae2b35db10b4d0a48d093d/data/utae_dynamicen.py#L137), then I get planet paths of 
```
('/10N/23E-188N/1311_3077_13/PF-SR/2018-01-01.tif',
 '/10N/23E-188N/1311_3077_13/PF-SR/2018-01-05.tif',
 '/10N/23E-188N/1311_3077_13/PF-SR/2018-01-10.tif',
 '/10N/23E-188N/1311_3077_13/PF-SR/2018-01-15.tif', 
'/10N/23E-188N/1311_3077_13/PF-SR/2018-01-20.tif', 
'/10N/23E-188N/1311_3077_13/PF-SR/2018-01-25.tif`)
```

And the corresponding label path is `'/labels/1311_3077_13_10N/Labels/Raster/10N-122W-40N-L3H-SR/10N-122W-40N-L3H-SR-2018_01_01.tif'`

Does that imply that the daily planet images are future steps of the actual annotation label or is this just a discrepency in the file naming. 

I expected in the weekly or daily case, that the inputs would be a time-series of images leading up to the label so in the above case the label would be `2018_02_01`.

**Comment by aysim** [comment]
We used the following images:
/10N/23E-188N/1311_3077_13/PF-SR/2018-01-01.tif,
/10N/23E-188N/1311_3077_13/PF-SR/2018-01-05.tif,
/10N/23E-188N/1311_3077_13/PF-SR/2018-01-10.tif,
/10N/23E-188N/1311_3077_13/PF-SR/2018-01-15.tif,
/10N/23E-188N/1311_3077_13/PF-SR/2018-01-20.tif,
/10N/23E-188N/1311_3077_13/PF-SR/2018-01-25.tif

along with the corresponding label:
/labels/1311_3077_13_10N/Labels/Raster/10N-122W-40N-L3H-SR/10N-122W-40N-L3H-SR-2018_01_01.tif

So it's more like sampling from the same month’s data pool rather than predicting a future time step. I totally get your point though — it’s more of a design choice in our case.


Regarding your earlier question:
"...The sentinel 1 tiff files have 8 bands, but I could not find which bands those actually are. Similarly for Sentinel 2, there are 12 bands but I could not find documentation or info which bands are included."

I’m not really the Sentinel person here, but I believe Andrés  can help with that — DLR prepared the data, so he should have the details. Might be worth sending him (andres.camerounzueta@dlr.de) a quick email (feel free to cc me too).
